[Core] Speculative Hold option for mod-taps: hold mods instantly while unsettled.

getreuer opened the pull request:
Speculative Hold applies mod-tap mods "speculatively" while the key is unsettled.

## Description

This is a core implementation of my community module [getreuer/speculative_hold](https://getreuer.info/posts/keyboards/speculative-hold/index.html). Speculative Hold makes mod-tap keys more responsive by applying the modifier instantly on keydown, before the tap-hold decision is made. This is especially useful for actions like <kbd>Shift</kbd>+click and <kbd>Ctrl</kbd>+scroll wheel with an external mouse, which can feel laggy with standard mod-taps.

The firmware holds the modifier speculatively. Once the key's behavior is settled:

* If held, the modifier remains active as expected until the key is released.
* If tapped, the speculative modifier is canceled just before the tapping keycode is sent.

Speculative Hold applies the modifier early but does not change the underlying tap-hold decision logic. Speculative Hold is compatible to use in combination with any other tap-hold options.

### Enable Speculative Hold

Once you have git checkout'd this PR, add the following to your `config.h`:

```c
#define SPECULATIVE_HOLD
```

### Configuration

By default, Speculative Hold applies to mod-taps using Shift, Ctrl, or Shift + Ctrl. You can override this behavior by defining the `get_speculative_hold()` callback in your keymap, for instance:

```cc
bool get_speculative_hold(uint16_t keycode, keyrecord_t* record) {
    switch (keycode) { // Enable speculative holding for these keys.
        case LCTL_T(KC_ESC):
        case LSFT_T(KC_Z):
        case RSFT_T(KC_SLSH):
            return true;
    }
    return false; // Disable otherwise.
}
```

A technicality is that some mods, such as GUI, may be bound to application actions when pressed by themselves, and Speculative Hold may falsely trigger such hotkeys. Define `DUMMY_MOD_NEUTRALIZER_KEYCODE` (and optionally `MODS_TO_NEUTRALIZE`) in your `config.h` to prevent this [as described here](https://docs.qmk.fm/tap_hold#retro-tapping).


## Implementation notes

Summary of how it works:

* A new kind of mods, `speculative_mods`, is implemented in action_util.

* An array `speculative_keys` is used as a queue to remember active speculatively-held keys.

* `process_speculative_press()` is called on the initial keydown event for a mod-tap key. It adds the mod-tap mods to `speculative_mods` and adds the keypos to `speculative_keys`.

* `process_speculative_hold()` is called when keys settle. It finds and removes the settled key from the `speculative_keys` queue and removes its mods from `speculative_mods`. 

  * If the key was settled as tapped, all following speculative keys in the queue are cleared as well.
  * If `DUMMY_MOD_NEUTRALIZER_KEYCODE` is defined, `neutralize_flashing_modifiers()` is called as needed.


Testing:

* An extensive suite of unit tests for Speculative Hold is added. I've included tests covering the behavior of Speculative Hold together with Flow Tap, Retro Tapping, and Retro Shift, where the potential for nontrivial interactions appears greater, since these features too customize tap-hold behavior at keydown or at the time when keys are settled.

* I've been running Speculative Hold on my keyboards for the past few weeks.


## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [x] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout (addition or update)
- [x] Documentation

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

IsaacElenbaas commented on docs/tap_hold.md:
```diff
@@ -745,6 +745,39 @@ Do not use `MOD_xxx` constants like `MOD_LSFT` or `MOD_RALT`, since they're 5-bi
+
+Speculative Hold makes mod-tap keys more responsive by applying the modifier instantly on keydown, before the tap-hold decision is made. This is especially useful for actions like Shift+Click with a mouse, which can feel laggy with standard mod-taps.
+
+The firmware holds the modifier speculatively. Once the key's behavior is settled:
```
I'm interested because I am reviewing, but also as someone intending to use the feature one of my immediate questions is "how does this interact with mod taps that are under layer taps?" What about when using two mod taps with tap hold configs that allow a roll to possibly mean two taps?

IsaacElenbaas replied:
At some point we just need a website to generate [this](https://web.archive.org/web/20230519045024/https://cdn.discordapp.com/attachments/451392568345755648/747388373781708930/tap-holds-qmk.pdf) with a given set of tap hold configs and _per_key functions :laughing:

So maybe the start of a thing we can add to the docs, if I am understanding you correctly:
> When one presses a layer-tap immediately before a key which has a mod-tap on the original layer and/or the layer activated by the layer-tap, Speculative Hold works as follows. "Key A" will refer to the possible mod-tap on the original layer, and "key B" will refer to the possible mod-tap on the layer-tap's layer.
> * The original layer is assumed, and if key A is a mod-tap, it is speculatively held
> * If rolling is possible given the tap-hold configurations and takes place (the layer-tap press was a tap event)
>   * Key A is processed in the usual Speculative Hold manner
> * Else
>   * When the layer-tap is evaluated (was released, its `TAPPING_TERM` elapses, etc.) nothing changes - key A is still speculatively held
>   * Once key B is evaluated (was released, its `TAPPING_TERM` elapses, etc.), key A's speculatively held modifier is cleared and key B is either tapped or has its modifier applied
> 
> In short, mod-taps "behind" a layer-tap are not speculatively held.

I feel like it might be possible to [clear a key's speculative storage entry and call `speculative_key_press` again] when it reaches the top of the waiting buffer to make speculative hold work optimally in all cases. That jank could always come in a future PR though, I love how clean this is. (aside - my initial comment said this feature scared me because I assumed it would be a ton of action_tapping logic)

IsaacElenbaas commented on quantum/action_tapping.c:
```diff
@@ -708,6 +734,110 @@ void waiting_buffer_scan_tap(void) {
+    if ((~(get_mods() | get_speculative_mods()) & mods) != 0 && get_speculative_hold(keycode, record)) {
+        add_speculative_mods(mods);
+        speculative_keys[num_speculative_keys] = (speculative_key_t){
+            .key  = record->event.key,
```
I'm still doing an initial pass so am not familiar with the code yet - but would storing the record->event.pos not be better? I imagine we only need this to release early (before the actual speculative-ness is collapsed and key events are sent) in which case checking the physical location seems ideal.

IsaacElenbaas replied:
I just fumbled some keys and was left with control stuck down. I can't manage to reproduce it, but it involved a `LCTL(KC_LEFT)` under a layer tap. Not sure if I hit my Retro Shifted `MT(MOD_LCTL, KC_SCLN)` on the main layer as well. This keymap has never stuck keys with a few years of use, though I did just flash those years' worth of QMK updates alongside this PR.

I'd definitely feel safer using `event->pos` if possible :slightly_smiling_face: just less chance for weird cases.

getreuer replied:
Thanks for the report and heads up, I'll see if I can repro that!

"`.key`" is perhaps a misnomer, it's a `keypos_t`, not a 16-bit keycode. I very much agree that keys need to be tracked based on matrix position, due to the layer ambiguity as described in the other comment.

IsaacElenbaas commented on quantum/action_tapping.c:
```diff
@@ -708,6 +734,110 @@ void waiting_buffer_scan_tap(void) {
+    }
+}
+
+void process_speculative_hold(keyrecord_t *record) {
```
Please add a comment noting that this is called at the top of process_record and in action_tapping.c to be alongside the other logic. I would also rename it to be more descriptive, personally - you are matching the naming of process_record_quantum functions, but flow_tap_update_last_event is already bailing on that for process_record and this is really confusing with there being a separate process_speculative_press in this file.

IsaacElenbaas commented on quantum/action_util.c:
```diff
@@ -398,6 +405,28 @@ void clear_weak_mods(void) {
 }
 
+#ifdef SPECULATIVE_HOLD
+uint8_t get_speculative_mods(void) {
+    return speculative_mods;
+}
+
+void set_speculative_mods(uint8_t mods) {
+    speculative_mods = mods;
+}
+
+void add_speculative_mods(uint8_t mods) {
+    speculative_mods |= mods;
+}
+
+void del_speculative_mods(uint8_t mods) {
+    speculative_mods &= ~mods;
+}
+
+void clear_speculative_mods(void) {
+    speculative_mods = 0;
+}
```
I suggest just doing these inline, unless there's something I'm missing. Breaking state by calling them from anything other than the implementation doesn't seem like it would be pretty and I don't think we can expose them.

IsaacElenbaas commented on quantum/action_tapping.c:
```diff
@@ -708,6 +734,110 @@ void waiting_buffer_scan_tap(void) {
+        }
+
+        del_speculative_mods(cleared_mods);
+        send_keyboard_report();
```
I think we need to sleep for `TAP_CODE_DELAY` here.

getreuer replied:
`neutralize_flashing_modifiers()` calls `tap_code()`, which includes a delay of `TAP_CODE_DELAY` between registering and unregistering the dummy neutralizing keycode. Should that be enough? Or is there another point here where a delay would help?

IsaacElenbaas replied:
This is the block for `Tap press event; cancel speculatively-held mod.` This function was called from `process_action`, which after we return will go through the usual `process_record_quantum` and such handling to press the key tap. If that is correct, we need to sleep after releasing the mod and sending a keyboard report so that those with the issues that `TAP_CODE_DELAY` is made to mitigate don't experience the speculative mod being applied to the tap.

IsaacElenbaas replied:
Maybe `TAP_CODE_DELAY` isn't intended for that case. . . but there isn't a better define AFAIK and I believe it was being used between every event of "shift down, tap key down, tap key up, shift up" for Auto Shift before I rewrote it.

IsaacElenbaas reviewed (approved):
Looks good to me, though I haven't touched QMK since Retro Shift so am not up-to-date on anything.

getreuer commented:
@IsaacElenbaas thank you very much for the thoughtful review comments and guidance!

Regarding the stuck Ctrl mod, I'll see I can repro it. I've added a test case involving LTs in https://github.com/qmk/qmk_firmware/pull/25572/commits/ccdc0a1e97aa1a601145529880222b4ad0159af8, which so far is green, but I'll try some variations under Retro Shift and hopefully trigger it.

getreuer commented:
@IsaacElenbaas thanks again for your great review comments.

I have added a thorough suite of tests involving layer-tap keys. While I haven't managed to repro the stuck mod key situation you mention, I did have an epiphany thinking about that: in the initial form of this PR, Speculative Hold could apply while other non-speculative events are buffered, potentially sending keys out of order. For instance if "`A ↓, B ↓, B ↑`" is in the waiting queue and then a mod-tap key is pressed, speculatively applying the mod immediately would apply the mod to `B`, even though `B` was released before the mod-tap was pressed.

To prevent such out of order keys, commit https://github.com/qmk/qmk_firmware/pull/25572/commits/d96bdf8102a6d63d067ed07b0738b97840566885 revises such that we don't do Speculative Hold when there are non-speculated buffered events.

I implemented that logic by storing `record.tap.speculated = true` in the `keyrecord_t`, then checking this field on `waiting_buffer` entries in `speculative_key_press()`. This spends one of the three unused reserved bits in the `tap_t` struct.

A good thing about this revision is that it is more restrictive about when Speculative Hold applies. Speculative Hold still triggers when it should, when a mod-tap or several mod-taps are held in isolation, but it no longer triggers when the mod-tap is pressed within a roll involving other unsettled, non-speculative tap-hold keys. My hope is this addresses whatever underlying issue caused that stuck mod.

IsaacElenbaas commented:
I had the stuck mod occur again, this time without keyspam. Still not sure what could have triggered it, I was in the zone and thought I used the modifier normally. Once again easy to unstick.

My board still has the original commit for this, so it may have been fixed since then, but my point is that if not, it's a very rare issue so not really a concern. (I'll try to reflash my board soon though)

tzarc commented:
> About 2 weeks ago, I added a couple mitigation measures in [5944b8b](https://github.com/qmk/qmk_firmware/commit/5944b8ba693a075497cc64c75ef4fab23c7d295f) and [ba6b90d](https://github.com/qmk/qmk_firmware/commit/ba6b90df18211410e8fb353920b5f0e3eea0c785) that potentially fix this.

If these resurface we can fix in a subsequent PR. For now, let's get some more eyes on this.
